# Secure Users API

We will secure most, but not all of the users-api. So set the auth to:

~~~javascript
    auth: {
      strategy: "jwt",
    },
~~~

On all routes, except:

- create
- authenticate

Otherwise we would have no endpoints for registration and authentication.

Because we are frequently calling deleteAllUsers in our auth-test and users-api-test, we need to be sure to re-register and authenticate whenever we do this. Here are revised versions which should pass now:

### user-api-test.js

~~~javascript
iimport { assert } from "chai";
import { assertSubset } from "../test-utils.js";
import { playtimeService } from "./playtime-service.js";
import { maggie, testUsers } from "../fixtures.js";
import { db } from "../../src/models/db.js";

const users = new Array(testUsers.length);

suite("User API tests", () => {
  setup(async () => {
    playtimeService.clearAuth();
    await playtimeService.createUser(maggie);
    await playtimeService.authenticate(maggie);
    await playtimeService.deleteAllUsers();
    for (let i = 0; i < testUsers.length; i += 1) {
      // eslint-disable-next-line no-await-in-loop
      users[0] = await playtimeService.createUser(testUsers[i]);
    }
    await playtimeService.createUser(maggie);
    await playtimeService.authenticate(maggie);
  });
  teardown(async () => {});

  test("create a user", async () => {
    const newUser = await playtimeService.createUser(maggie);
    assertSubset(maggie, newUser);
    assert.isDefined(newUser._id);
  });

  test("delete all user", async () => {
    let returnedUsers = await playtimeService.getAllUsers();
    assert.equal(returnedUsers.length, 4);
    await playtimeService.deleteAllUsers();
    await playtimeService.createUser(maggie);
    await playtimeService.authenticate(maggie);
    returnedUsers = await playtimeService.getAllUsers();
    assert.equal(returnedUsers.length, 1);
  });

  test("get a user", async () => {
    const returnedUser = await playtimeService.getUser(users[0]._id);
    assert.deepEqual(users[0], returnedUser);
  });

  test("get a user - bad id", async () => {
    try {
      const returnedUser = await playtimeService.getUser("1234");
      assert.fail("Should not return a response");
    } catch (error) {
      assert(error.response.data.message === "No User with this id");
      assert.equal(error.response.data.statusCode, 503);
    }
  });

  test("get a user - deleted user", async () => {
    await playtimeService.deleteAllUsers();
    await playtimeService.createUser(maggie);
    await playtimeService.authenticate(maggie);
    try {
      const returnedUser = await playtimeService.getUser(users[0]._id);
      assert.fail("Should not return a response");
    } catch (error) {
      assert(error.response.data.message === "No User with this id");
      assert.equal(error.response.data.statusCode, 404);
    }
  });
});

~~~

Verify that all tests are now passing.